rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,291 @@
+`search` service.)
+
+Applications query data from Elasticsearch on behalf of the end user in a variety of ways:
+
+- Using `SearchSource` (our higher-level abstracted search API, used by Maps, Lens, and most visualizations, exposed by
+the `data` plugin `search` service)
+- Using `search` (our lower-level search API that accepts raw Elasticsearch query DSL, used by Vega and Timelion,
+exposed by the `data` plugin `search` service)
+- Exposing a route that bypasses our `search` services and queries Elasticsearch directly, then manipulates the data
+before returning a response (as in the case of TSVB and most solutions)
+
+The proposed APIs provide the tools necessary to run long search requests, view their status, cancel incomplete search
+requests, and retrieve results, regardless of the mechanism used to query Elasticsearch.
+
+## Background session service
+
+The primary service which will facilitate background sessions is the server-side `session` service.
+
+When a `search` request is handled, and it contains a `sessionId`, the service will query Elasticsearch to see if there
+is already a saved object for that session. If there isn't, then we will forward the search request to Elasticsearch and
+receive an ID specific to that search request. We store that ID in memory, keyed off of a hash of the request parameters
+and the current username.
+
+We will then periodically query Elasticsearch to see if there is a saved object for that session (meaning the user has
+elected to continue the requests in the background). If we find a saved object, then we will update it to contain the
+search ID keyed off of the request hash, and remove it from memory.
+
+Then, in the future when the same `sessionId` is sent with identical request parameters, we will retrieve the saved
+object. Instead of sending all of the request parameters to Elasticsearch (and restarting the search), we will look up
+the search ID, and send that to Elasticsearch (resuming the search). If that specific search request has completed in
+Elasticsearch, the results will be available immediately.
+
+By default, results are stored in Elasticsearch for 5 days. We may consider adding an advanced setting that will allow
+overriding this default.
+
+The background session service will also use the task manager to periodically monitor the status of incomplete
+background sessions. It will query the list of all incomplete sessions, and check the status of each search that is
+executing. If the search requests are all complete, it will update the corresponding saved object to have a `status` of
+`complete`.
+
+## Server APIs
+
+The following APIs are exposed by the server-side `data` plugin start `search` service:
+
+```ts
+session.getAll()
+```
+
+Retrieves all of the background session saved objects for the current user. (We can filter it to the current user
+because when the saved object is created, we store a hash of the current user's username inside the object.)
```

Comment:
Just the username isn't sufficient for these to be "per user"... @elastic/kibana-security if they were to store the username AND the realm, would that be sufficient?

Reply:
> @elastic/kibana-security if they were to store the username AND the realm, would that be sufficient?

The most unique thing we can get right now is `username:realm-name:realm-type` (or `username:auth-provider-name:auth-provider-type`). There is a risk that realm with the same name and type will be reconfigured to work with another IdP/user-store, but we consider this risk as low.

Reply:
I know there is an API to get the current username, but is there a client-side API to get the realm and realm type?